Speed up the learning time of EM for large dataset

reeingal opened the pull request:
The `_compute_weights()` method of `ExpectationMaximization` will
traverse all unique data in the dataset use `for` loop.

if the count of unique data is large, the `_compute_weights()` method
will take too much time in `for` loop.

I make a try to use `joblib.Parallel` for paralleling, and find that it
saves 70% to 80% time on my dataset with 9994 unique records and
10 columns. The number of latent variables in my test are 3.

### Your checklist for this pull request
Please review the [guidelines for contributing](CONTRIBUTING.md) to this repository.

- [ ] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). Don't request your master!
- [ ] Make sure you are making a pull request against the **dev branch** (left side). Also you should start *your branch* off *our dev*.
- [ ] Check the commit's or even all commits' message styles matches our requested structure.

### Issue number(s) that this pull request fixes
- Fixes #1526

### List of changes to the codebase in this pull request
- change the `for` loop in  `_compute_weights()` method of `ExpectationMaximization` to parallel running expected to improve executing time.

reeingal committed:
Speed up the learning time of EM for large dataset

The `_compute_weights()` method of ExpectationMaximization will
traverse all unique data in the dataset use `for` loop.

if the count of unique data is large, the `_compute_weights()` method
will take too much time in `for` loop.

I make a try to use `joblib.Parallel` for paralleling, and find that it
saves 70% to 80% time on my dataset with 9994 unique records and
10 columns. The number of latent variables in my test are 3.

ankurankan commented:
@reeingal Could you run `black .` in the pgmpy directory to fix the formatting and commit the changes? The tests are failing because of that.

reeingal committed:
use black to format the code

reeingal committed:
Fix the wrong offset pass to parallel

ankurankan commented on pgmpy/estimators/EM.py:
```diff
@@ -105,6 +100,26 @@ def _compute_weights(self, latent_card):
+
+        batch_size = 1000
+
+        cache = Parallel(n_jobs=-1)(
```
@reeingal `n_jobs` here should be taken as a method argument. The `get_parameters` method has a `n_jobs` argument which it can pass to `_compute_weights` when called.

ankurankan commented on pgmpy/estimators/EM.py:
```diff
@@ -105,6 +100,26 @@ def _compute_weights(self, latent_card):
+        data_unique = self.data.drop_duplicates()
+        n_counts = self.data.groupby(list(self.data.columns)).size().to_dict()
+
+        batch_size = 1000
```
I would suggest having `batch_size` also as an argument with the default value of 1000.

ankurankan reviewed (changes requested):
@reeingal Seems to be working fine for me on realistic datasets. I tested it on the alarm and munin1 network: https://www.bnlearn.com/bnrepository/ with up to 1e5 samples and seems to be much faster in most of the cases. Though the run time is higher when the dataset size is less than or around 1e3. I would suggest adding a note in the `get_parameters` docstring to use n_jobs=1 when dataset size is less than 1000.

reeingal committed:
Take n_jobs as a argument of _compute_weights

Take `n_jobs` as a argument of `compute_weights()` to control the number
of paralleling processes. It uses the `n_jobs` of `get_parameters()`.
Suggest to use `n_jobs=1` when dataset
size less than 1000.

reeingal committed:
Add batch_size as a argument of get_parameters

Add `batch_size` as a argument of `get_parameters()`, and pass it to
`_compute_weights()` to control  number of data used to compute weights
in a batch.
